[Dione Goh] iP

dionegoh opened the pull request:
# PikaBot ⚡ ⚡ 

> "Pikapika!"  [-Pikachu](https://www.pokemon.com/us/pokedex/pikachu)

PikaBot is a task manager that helps you to manage your tasks efficiently! It is: 
- `text-based`
- `easy to use`

 ## **Features** 😛 
- [x] Adding tasks (todos, events, deadlines)
- [x] Deleting tasks
- [x] Listing tasks
- [x] Searching for tasks
- [x] Marking tasks as done

Here's a snippet of `code` in PikaBot to show how PikaBot processes inputs 📝 !
```java
while (!strInputArr[0].equals("bye")) {
      Command command = Parser.parseCommand(strInputArr);
      command.execute(taskList, storage);
      input = sc.nextLine();
     strInputArr = input.split(" ", 2);
}
```
## _How to download PikaBot?_
1. Download PikaBot.jar [here](https://github.com/dionegoh/ip/releases/tag/A-Jar)!
2. Run PikaBot.jar
3. Start adding your tasks! 👍

Commit:
Add Gradle support

dionegoh committed:
Completed Level-1

dionegoh committed:
Completed Level-2

dionegoh committed:
Completed Level-3

dionegoh committed:
Completed Level-4

dionegoh committed:
Tested A-TextUiTesting

dionegoh committed:
Completed Level-5

dionegoh committed:
Completed Level-6

dionegoh committed:
Complete Level-7

dionegoh committed:
Complete Level-8

dionegoh committed:
Merge branch 'branch-Level-7'

dionegoh committed:
Merge with branch-Level-8

dionegoh committed:
Use more OOP

dionegoh committed:
Organize into Packages

dionegoh committed:
Add JUnit Tests

dionegoh committed:
Add JUnit Tests

dionegoh committed:
Modify runtest.sh after refactoring into packages

dionegoh committed:
Add JavaDoc comments

dionegoh committed:
Tweak the code to comply with a coding standard

dionegoh committed:
Merge with branch-A-JavaDoc

dionegoh committed:
Merge with branch-A-CodingStandard

dionegoh committed:
New functionality to find a task by searching for a keyword

dionegoh committed:
Merge with branch-Level-9

dionegoh committed:
Modify README.md

dionegoh committed:
Reformat code to follow coding standard

ChanWeiJie commented on src/main/java/pikabot/PikaBot.java:
```diff
@@ -0,0 +1,53 @@
+    public static void main(String[] args) {
+
+        Scanner sc = new Scanner(System.in);
+        Storage storage = new Storage("data/tasks.txt");
```
Perhaps a better way would be to use the <code>static String FILEPATH</code> on line 18 as the parameter for the call to Storage? 🤔

ChanWeiJie commented on src/main/java/pikabot/Parser.java:
```diff
@@ -0,0 +1,152 @@
+            throw new FindException();
+        }
+    }
+}
```
I like how you seperated each parsing of the respective Task into individual methods ! 👍

ChanWeiJie commented on src/main/java/pikabot/Storage.java:
```diff
@@ -0,0 +1,136 @@
+            String taskStr = sc.nextLine();
+            char taskType = taskStr.charAt(1);
+            boolean isTaskDone = (taskStr.charAt(4) == 'X');
+            taskStr = taskStr.substring(7);
```
Perhaps declaring a <code>private final String TASK_DESCRIPTION</code>  for the use of  "7" here ?

ChanWeiJie commented on src/main/java/pikabot/Storage.java:
```diff
@@ -0,0 +1,136 @@
+     * a directory rather than a regular file, does not exist
+     * but cannot be created, or cannot be opened for any other reason.
+     */
+    public void appendToFile(Task taskToAdd) throws IOException{
```
I like how clean this bit of code is! But, with regards to the coding standard, there should be a space after <code> IOException </code>, other than that good job ! 😄

ChanWeiJie reviewed (commented):
LGTM! Just some nix to fix.

dionegoh committed:
Add more exceptions for invalid commands

dionegoh committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

dionegoh committed:
Automate project builds using Gradle

dionegoh committed:
Merge branch branch-A-Gradle

dionegoh committed:
Use checkStyle

dionegoh committed:
Fix more errors with checkStyle

dionegoh committed:
Add GUI - in progress

dionegoh committed:
Complete JavaFX tutorial part 3

geetengtan commented on src/main/java/pikabot/TaskList.java:
```diff
@@ -0,0 +1,104 @@
+ */
+public class TaskList {
+
+    protected ArrayList<Task> taskList;
```
Might be confusing for "TaskList" class to have an attribute "taskList"? "taskList" can be changed to "tasks"

geetengtan commented on src/main/java/pikabot/Ui.java:
```diff
@@ -0,0 +1,144 @@
+     * Prints out welcome message when PikaBot starts running.
+     */
+    public static void printWelcomeText() {
+        System.out.println(INDENTATION + LINE + "\n" + INDENTATION + "Hello! I'm PikaBot"
```
Could use more line breaks to improve readability

geetengtan commented on src/main/java/pikabot/Ui.java:
```diff
@@ -0,0 +1,144 @@
+            taskNumber++;
+        }
+
+        System.out.println(INDENTATION + LINE);
```
Great use of linebreaks to separate different sections of the method

geetengtan commented on src/main/java/pikabot/Ui.java:
```diff
@@ -0,0 +1,144 @@
+                + INDENTATION + "Got it. I've added this task:" + "\n"
+                + INDENTATION + "  " + task + "\n"
+                + INDENTATION + "Now you have " + taskList.noOfTasks() + " tasks in the list."
+                + "\n" + INDENTATION + LINE);
```
Might be nit picking here, but if the "\n" can be on the previous line, the readability will be 💯

geetengtan replied:
Provided it doesn't exceed the character limit per line

geetengtan commented on src/test/java/pikabot/ParserTest.java:
```diff
@@ -0,0 +1,38 @@
+        }
+    }
+
+}
```
Detailed method naming for test methods 🥇

geetengtan reviewed (commented):
LGTM, love pikachu and love pikabot even more! 😃

dionegoh committed:
Complete add GUI

dionegoh committed:
Complete add GUI

ivorcmx commented on src/main/java/pikabot/Parser.java:
```diff
@@ -0,0 +1,170 @@
+     * @param inputArr Array containing input string from user.
+     * @return Command.
+     */
+    public static Command parseCommand(String[] inputArr) {
```
Readability of this block of code is very good

ivorcmx commented on src/main/java/pikabot/Parser.java:
```diff
@@ -0,0 +1,170 @@
+     * @return Todo command.
+     * @throws TodoException If input is invalid.
+     */
+    public static Todo parseTodo(String[] todoArray) throws TodoException {
```
Nice naming convention used to describe the string array

ivorcmx commented on src/main/java/pikabot/TaskList.java:
```diff
@@ -0,0 +1,104 @@
+     *
+     * @param taskNumber Index of task to be marked as done, one-based indexing.
+     */
+    public void markTaskAsDone(int taskNumber) {
```
Nice method naming to explain clearly what the method does

ivorcmx commented on src/main/java/pikabot/command/MarkCommand.java:
```diff
@@ -0,0 +1,48 @@
+    public String execute(TaskList taskList, Storage storage) {
+        try {
+            Parser.parseIntegerCommand(markCommand);
+            int taskToMark = Integer.parseInt(markCommand[1]);
```
Good variable naming for clarity

ivorcmx commented on src/test/java/pikabot/ParserTest.java:
```diff
@@ -0,0 +1,38 @@
+public class ParserTest {
+
+    @Test
+    public void todo_withNoDescription_exceptionThrown() {
```
Good method naming to suit what it does!

ivorcmx reviewed (commented):
Interesting and well done project with little to no code quality issues!

dionegoh committed:
Add text wrap

dionegoh committed:
Modify values for Shadow plugin

dionegoh committed:
Add assertions

dionegoh committed:
Refactor to improve code quality

To improve code quality, some portions of the code require refactoring.

- There are some methods with similar behaviours causing code
duplication.

Extracting common behaviour into a single method allows us to adhere to
D-R-Y (don't repeat yourself) principle. Let's find multiple methods
with common behaviours and extract them into a single method.

- There are lines of dead code.

Let's delete the lines of dead code for safe practice.

- There are some "if-else" constructs where the final "else" refers to
the final option.

Using default branch for intended default action and not just to
execute last option improves code quality. Let's change the final
"else" to mean "everything else", and not the final option.

dionegoh committed:
Add friendlier syntax for commands

Making the command syntax more flexible is beneficial for users who do
not have the time to key in the full command.

Let's provide shorter aliases for key commands.

dionegoh committed:
Merge branch 'branch-A-Assertions'

dionegoh committed:
Resolve merge conflict

dionegoh committed:
Resolve merge conflict

dionegoh committed:
Merge branch 'branch-A-CodeQuality'

To improve code quality, some portions of the code require refactoring.

- There are some methods with similar behaviours causing code
duplication.
Extracting common behaviour into a single method allows us to
adhere to D-R-Y (don't repeat yourself) principle. Let's find
multiple methods with common behaviours and extract them into a
single method.

- There are lines of dead code.
Let's delete the lines of dead code for safe practice.

- There are some "if-else" constructs where the final "else"
refers to the final option.
Using default branch for intended default action and not just to
execute last option improves code quality. Let's change the final
"else" to mean "everything else", and not the final option.

dionegoh committed:
Merge branch 'master' into branch-BCD-extension

dionegoh committed:
Merge branch 'branch-BCD-Extension'

Make syntax friendlier.

Making the command syntax more flexible is beneficial for users
who do not have the time to key in the full command.

Let's provide shorter aliases for key commands.

dionegoh committed:
Set up continuous integration

dionegoh committed:
Set up continuous integration

dionegoh committed:
Add better GUI

Tweak GUI to make it more more polished.

dionegoh committed:
Add representative screenshot

dionegoh committed:
Add user guide

dionegoh committed:
Fix alignment issues in user guide

dionegoh committed:
Fix alignment issues in user guide

dionegoh committed:
Add toLowerCase for bye command

dionegoh committed:
Fix bug in build.gradle
